scalanet/src/io/iohk/scalanet/peergroup/kademlia/KPeerGroup.scala

```diff
@@ -0,0 +1,161 @@
+package io.iohk.scalanet.peergroup.kademlia
+
+import java.net.{InetAddress, InetSocketAddress}
+
+import io.iohk.scalanet.peergroup.PeerGroup.ServerEvent.ChannelCreated
+import io.iohk.scalanet.peergroup.PeerGroup.{ChannelSetupException, HandshakeException, ServerEvent}
+import io.iohk.scalanet.peergroup.kademlia.KPeerGroup.{ChannelImpl, NodeRecord, UnderlyingChannel}
+import io.iohk.scalanet.peergroup.{Channel, InetMultiAddress, PeerGroup}
+import monix.eval.Task
+import monix.execution.Scheduler
+import monix.reactive.Observable
+import monix.reactive.subjects.Subject
+import org.slf4j.{Logger, LoggerFactory}
+import scodec.bits.BitVector
+
+// PeerGroup using node records for addressing.
+// These node records are derived from Ethereum node records (https://eips.ethereum.org/EIPS/eip-778)
+class KPeerGroup[M](
+    val kRouter: KRouter[NodeRecord],
+    val server: Subject[ServerEvent[BitVector, M], ServerEvent[BitVector, M]],
+    underlyingPeerGroup: PeerGroup[InetMultiAddress, Either[NodeRecord, M]]
+)(implicit scheduler: Scheduler)
+    extends PeerGroup[BitVector, M] {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  override def processAddress: BitVector = kRouter.config.nodeId
+
+  override def initialize(): Task[Unit] = {
+    // The protocol defined here requires that a peer
+    // connects then sends it node record as the first message
+    // (it could send earlier messages but they will be ignored)
+    // This node will then create a channel and reply with its
+    // own node record.
+    underlyingPeerGroup
+      .server()
+      .collect(ChannelCreated.collector)
+      .mergeMap(
+        channel =>
+          channel.in.collect {
+            case Left(nodeRecord) =>
+              acceptNodeRecord(channel, nodeRecord)
+          }
+      )
+      .subscribe()
+
+    Task.unit
```

Comment:
Two comments:

First, I believe this implementation violates the semantics of the interface.
When a method returns a `Task`, the interpretation I have is that the computation has not started and won't start until I call `runAsync` (or equivalent method), this carries that no side effect should happen before "running" the Task. Is my interpretation wrong?
In this implementation, the underlying peergroup is actually called before I decide to execute the Task.

Another interesting thing is that the underlying peergroup is never initialized. I think this should never actually receive a message if this is correct.

Reply:
Good point about Task. I've updated.

Re initialization. The current, unwritten convention is that a peer group does not own its underlying peer group, therefore does not initialize it or shut it down. This convention came about when peer groups supported messages of _any_ type. This is no longer true, peer group message types are fixed when the peer group is instantiated and the the type signatures of underlying peer groups are often specialized to their parent's. This argues, then, that we should write a story which changes the convention, to where `initialize` on a peer group should initialize its underlying peer group (and so until the terminal peer group) and likewise for shutdown. As it stands today, we do not yet have this, so it is assumed that the user code will initialize both parent and child peer groups (and if it doesn't it won't work). Hope that makes sense.